File: english-learning/src/api/english.ts
/** /api/v1/english 相关接口的封装 */
import { request } from './http'
import type {
  ActivityReport,
  Article,
  ArticleItem,
  Category,
  CheckinStats,
  CollectResult,
  CollectionItem,
  CompleteResult,
  DailyReadingToday,
  DailySummary,
  MemoryRow,
  Note,
  ReadingArchiveItem,
  ReadingQuiz,
  RegisterResponse,
  RewardOverview,
  SrsSessionState,
  SrsSettings,
  SrsState,
  SrsStats,
  ShopItem,
  StudyStats,
  TestQuestions,
  TokenPair,
  UserProfile,
  Word,
  WordBook,
  WordbookItem,
} from '@/types'

// ── 认证 ─────────────────────────────────────────────────────────

export function login(username: string, password: string): Promise<TokenPair> {
  return request<TokenPair>('/tenant/auth/login_with_tenant', {
    method: 'POST',
    auth: false,
    query: { tenant_code: 'english' },
    body: { username, password },
  })
}

export function register(data: {
  username: string
  email: string
  password: string
}): Promise<RegisterResponse> {
  return request<RegisterResponse>('/english/auth/register', {
    method: 'POST',
    auth: false,
    body: data,
  })
}

export function refreshToken(refresh_token: string): Promise<TokenPair> {
  return request<TokenPair>('/english/auth/refresh', {
    method: 'POST',
    auth: false,
    body: { refresh_token },
  })
}

export function getProfile(): Promise<UserProfile> {
  return request<UserProfile>('/english/auth/profile')
}

export function updateProfile(data: { nickname?: string; avatar?: string }): Promise<UserProfile> {
  return request<UserProfile>('/english/auth/profile', { method: 'PUT', body: data })
}

// ── 单词 / 词书 ──────────────────────────────────────────────────

export function listWords(params?: { book_id?: number; level?: string; skip?: number; limit?: number }): Promise<Word[]> {
  return request<Word[]>('/english/words', { query: params })
}

export function lookupWord(word: string): Promise<{ word: string; phonetic: string | null; definition: string; pos: string | null }> {
  return request('/english/words/lookup', { query: { word } })
}

export function listBooks(): Promise<WordBook[]> {
  return request<WordBook[]>('/english/srs/books', { auth: false })
}

// ── SRS 背单词 ────────────────────────────────────────────────────

export function getSrsState(book_id?: number): Promise<SrsState> {
  return request<SrsState>('/english/srs/state', { query: { book_id } })
}

export function saveSrsState(data: {
  book_id: number
  memory: MemoryRow[]
  session: SrsSessionState | null
}): Promise<{ saved: boolean }> {
  return request<{ saved: boolean }>('/english/srs/state', { method: 'POST', body: data })
}

export function onboarding(data: {
  target: string
  book_id: number
  daily_new_words: number
  pronunciation: string
  autoplay: boolean
}): Promise<SrsSettings> {
  return request<SrsSettings>('/english/srs/onboarding', { method: 'POST', body: data })
}

export function completeDay(data: {
  book_id: number
  study_date: string
  review_count: number
  new_count: number
  wrong_count: number
}): Promise<CompleteResult> {
  return request<CompleteResult>('/english/srs/complete', { method: 'POST', body: data })
}

export function syncGuest(data: { book_id: number; memory: MemoryRow[]; wordbook: number[] }): Promise<{ memory_merged: number; wordbook_merged: number }> {
  return request<{ memory_merged: number; wordbook_merged: number }>('/english/srs/sync', { method: 'POST', body: data })
}

export function resetBook(book_id: number): Promise<{ reset: boolean }> {
  return request<{ reset: boolean }>('/english/srs/reset', { method: 'POST', body: { book_id } })
}

export function getSrsStats(): Promise<SrsStats> {
  return request<SrsStats>('/english/srs/stats')
}

export function setWordTag(book_id: number, word_id: number, tag: string | null): Promise<{ word_id: number; tag: string | null }> {
  return request<{ word_id: number; tag: string | null }>('/english/srs/tag', { method: 'PUT', body: { book_id, word_id, tag } })
}

export function getCategories(book_id: number): Promise<Category[]> {
  return request<Category[]>('/english/srs/categories', { query: { book_id } })
}

export function getTestQuestions(params: {
  book_id: number
  module: 'choice' | 'fill'
  question_type: 'a' | 'b' | 'c' | 'd' | 'e'
  mode: 'today' | 'book' | 'wordbook' | 'reading_new'
  count: number
}): Promise<TestQuestions> {
  return request<TestQuestions>('/english/test/questions', { query: params })
}

// ── 每日一读 ─────────────────────────────────────────────────────

export function getDailyReadingToday(): Promise<DailyReadingToday> {
  return request<DailyReadingToday>('/english/daily-reading/today')
}

export function setReadingLevel(mode: 'auto' | 'manual', level?: 'basic' | 'cet4' | 'advanced'): Promise<{ saved: boolean; mode: string; level: string | null }> {
  return request('/english/daily-reading/level', { method: 'PUT', body: { mode, level } })
}

export function getReadingQuiz(article_id: number): Promise<ReadingQuiz> {
  return request<ReadingQuiz>('/english/daily-reading/quiz', { query: { article_id } })
}

export function submitReadingQuiz(
  article_id: number,
  answers: { word: string; type: string; correct: boolean; definition?: string }[],
): Promise<{ saved: boolean; correct: number; total: number }> {
  return request('/english/daily-reading/quiz', { method: 'POST', body: { article_id, answers } })
}

export function collectReadingWord(article_id: number, word: string, source: string, definition?: string): Promise<{ status: string; reason?: string; word_id?: number }> {
  return request('/english/daily-reading/words', { method: 'POST', body: { article_id, word, source, definition } })
}

export function setReadingBlacklist(word: string, blacklisted: boolean): Promise<{ saved: boolean; blacklisted: boolean }> {
  return request('/english/daily-reading/words/blacklist', { method: 'PUT', body: { word, blacklisted } })
}

export function completeDailyReading(article_id: number, duration_sec: number): Promise<{ saved: boolean; status: string }> {
  return request('/english/daily-reading/complete', { method: 'POST', body: { article_id, duration_sec } })
}

export function getReadingArchive(): Promise<ReadingArchiveItem[]> {
  return request<ReadingArchiveItem[]>('/english/daily-reading/archive')
}

// ── AI 每日学习总结 ──────────────────────────────────────────────

export function reportActivity(data: ActivityReport): Promise<{ saved: boolean }> {
  return request<{ saved: boolean }>('/english/activity', { method: 'POST', body: data })
}

export function getTodayActivity(): Promise<{ has_activity: boolean }> {
  return request<{ has_activity: boolean }>('/english/activity/today')
}

export function getDailySummary(): Promise<DailySummary> {
  return request<DailySummary>('/english/daily-summary')
}

export function generateDailySummary(): Promise<DailySummary> {
  return request<DailySummary>('/english/daily-summary/generate', { method: 'POST' })
}

// ── 生词本 ────────────────────────────────────────────────────────

export function listWordbook(book_id?: number): Promise<WordbookItem[]> {
  return request<WordbookItem[]>('/english/wordbook', { query: { book_id } })
}

export function addWordbook(word_id: number, book_id: number): Promise<WordbookItem> {
  return request<WordbookItem>('/english/wordbook', { method: 'POST', body: { word_id, book_id } })
}

export function removeWordbook(id: number): Promise<{ deleted: boolean }> {
  return request<{ deleted: boolean }>(`/english/wordbook/${id}`, { method: 'DELETE' })
}

export function clearWordbook(book_id: number): Promise<{ deleted: boolean }> {
  return request<{ deleted: boolean }>('/english/wordbook', { method: 'DELETE', query: { book_id } })
}

// ── 阅读 ─────────────────────────────────────────────────────────

export function listArticles(): Promise<ArticleItem[]> {
  return request<ArticleItem[]>('/english/reading/articles')
}

export function getArticle(id: number): Promise<Article> {
  return request<Article>(`/english/reading/articles/${id}`)
}

export function listNotes(): Promise<Note[]> {
  return request<Note[]>('/english/reading/notes')
}

export function createNote(article_id: number, content: string): Promise<Note> {
  return request<Note>('/english/reading/notes', { method: 'POST', body: { article_id, content } })
}

export function updateNote(id: number, content: string): Promise<Note> {
  return request<Note>(`/english/reading/notes/${id}`, { method: 'PUT', body: { content } })
}

export function deleteNote(id: number): Promise<{ deleted: boolean }> {
  return request<{ deleted: boolean }>(`/english/reading/notes/${id}`, { method: 'DELETE' })
}

// ── 收藏 ─────────────────────────────────────────────────────────

export function listCollections(item_type?: string): Promise<CollectionItem[]> {
  return request<CollectionItem[]>('/english/collections', { query: { item_type } })
}

export function addCollection(item_type: string, item_id: number): Promise<CollectionItem> {
  return request<CollectionItem>('/english/collections', {
    method: 'POST',
    body: { item_type, item_id },
  })
}

export function removeCollection(id: number): Promise<{ deleted: boolean }> {
  return request<{ deleted: boolean }>(`/english/collections/${id}`, { method: 'DELETE' })
}

// ── 打卡 / 统计 ──────────────────────────────────────────────────

export function checkin(): Promise<CheckinStats> {
  return request<CheckinStats>('/english/checkin', { method: 'POST' })
}

export function getCheckinStats(): Promise<CheckinStats> {
  return request<CheckinStats>('/english/checkin/stats')
}

export function getStudyStats(): Promise<StudyStats> {
  return request<StudyStats>('/english/study/stats')
}

// ── 奖励系统 ─────────────────────────────────────────────────────

export function getRewardsOverview(): Promise<RewardOverview> {
  return request<RewardOverview>('/rewards/overview')
}

export function getRewardsShop(): Promise<ShopItem[]> {
  return request<ShopItem[]>('/rewards/shop')
}

export function collectRewards(): Promise<CollectResult> {
  return request<CollectResult>('/rewards/collect', { method: 'POST' })
}

export function redeemReward(itemKey: string): Promise<ShopItem> {
  return request<ShopItem>('/rewards/redeem', { method: 'POST', body: { item_key: itemKey } })
}

export function equipReward(
  itemKey: string | null,
): Promise<{ equipped_title: string | null; equipped_decor: string | null }> {
  return request<{ equipped_title: string | null; equipped_decor: string | null }>('/rewards/equip', {
    method: 'POST',
    body: { item_key: itemKey },
  })
}

File: backend/app/api/v1/english.py
"""英语学习 API 路由

认证说明：
- 注册/刷新为公开接口；
- /srs/books、/words、/reading 等内容接口游客可读（OptionalUserDep）；
- 其余接口均需 Bearer Token（UserDep）；
- 登录复用 /api/v1/tenant/auth/login_with_tenant?tenant_code=english
- 统一响应包裹：{code, data, message}
"""
from typing import Any

from fastapi import APIRouter

from app.core.dependencies import DbDep, OptionalUserDep, UserDep
from app.core.response import UnifiedResponse
from app.services.english.schemas import (
    ActivityIn,
    ArticleListItemOut,
    ArticleOut,
    CategoryOut,
    CheckinStatsOut,
    CollectOut,
    CollectionCreate,
    CollectionItemOut,
    CompleteIn,
    CompleteOut,
    DailyReadingRecordOut,
    DailyReadingTodayOut,
    DailySummaryOut,
    EquipIn,
    NoteCreate,
    NoteOut,
    NoteUpdate,
    OnboardingIn,
    ProfileOut,
    ProfileUpdate,
    RedeemIn,
    ReadingArchiveItemOut,
    ReadingBlacklistIn,
    ReadingCompleteIn,
    ReadingLevelIn,
    ReadingQuizSubmitIn,
    ReadingWordCollectIn,
    RefreshRequest,

    RegisterRequest,
    ResetIn,
    RewardOverviewOut,
    SetTagIn,
    ShopItemOut,
    SrsSettingsOut,
    SrsStateIn,
    SrsStateOut,
    StudyStatsOut,
    SyncIn,
    SyncOut,
    TestQuestionsOut,
    TodayActivityOut,
    TokenPair,
    WordBookOut,
    WordOut,
    WordbookAddIn,
    WordbookItemOut,
)
from app.services.english.service import EnglishService
from app.services.english.srs_service import SrsService
from app.services.english.test_service import TestService
from app.services.english.daily_summary_service import DailySummaryService
from app.services.english.daily_reading_service import DailyReadingService
from app.services.english.reward_service import RewardService

router = APIRouter(prefix="/english", tags=["英语学习"])


def _service(db: DbDep) -> EnglishService:
    return EnglishService(db)


def _srs(db: DbDep) -> SrsService:
    return SrsService(db)


def _reading(db: DbDep) -> DailyReadingService:
    return DailyReadingService(db)


# ── 认证 ─────────────────────────────────────────────────────────────


@router.post("/auth/register", response_model=UnifiedResponse[Any])
def register(data: RegisterRequest, db: DbDep) -> UnifiedResponse[Any]:
    service = _service(db)
    user, access, refresh = service.register(data.username, data.email, data.password)
    return UnifiedResponse.success(
        data={
            "user": ProfileOut.model_validate(user).model_dump(),
            "access_token": access,
            "refresh_token": refresh,
            "token_type": "bearer",
        },
        message="注册成功",
    )


@router.post("/auth/refresh", response_model=UnifiedResponse[TokenPair])
def refresh(data: RefreshRequest, db: DbDep) -> UnifiedResponse[TokenPair]:
    access, refresh_token = _service(db).refresh(data.refresh_token)
    return UnifiedResponse.success(data=TokenPair(access_token=access, refresh_token=refresh_token))


@router.get("/auth/profile", response_model=UnifiedResponse[ProfileOut])
def get_profile(user: UserDep, db: DbDep) -> UnifiedResponse[ProfileOut]:
    return UnifiedResponse.success(data=_service(db).get_profile(user))


@router.put("/auth/profile", response_model=UnifiedResponse[ProfileOut])
def update_profile(data: ProfileUpdate, user: UserDep, db: DbDep) -> UnifiedResponse[ProfileOut]:
    return UnifiedResponse.success(data=_service(db).update_profile(user, data.nickname, data.avatar))


# ── 单词 / 词书 ──────────────────────────────────────────────────────


@router.get("/words", response_model=UnifiedResponse[list[WordOut]])
def list_words(
    user: OptionalUserDep,
    db: DbDep,
    book_id: int | None = None,
    level: str | None = None,
    skip: int = 0,
    limit: int = 100,
) -> UnifiedResponse[list[WordOut]]:
    return UnifiedResponse.success(data=_service(db).list_words(user, book_id, level, skip, limit))


@router.get("/words/lookup", response_model=UnifiedResponse[dict])
def word_lookup(word: str, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_service(db).lookup_word(user, word))


# ── SRS 背单词 ───────────────────────────────────────────────────────


@router.get("/srs/books", response_model=UnifiedResponse[list[WordBookOut]])
def srs_books(user: OptionalUserDep, db: DbDep) -> UnifiedResponse[list[WordBookOut]]:
    if user:
        tenant_id = user.tenant_id
    else:
        from app.services.english.service import get_english_tenant

        tenant_id = get_english_tenant(db).id  # 游客回退到 english 租户
    return UnifiedResponse.success(data=_srs(db).get_books(tenant_id))


@router.get("/srs/state", response_model=UnifiedResponse[SrsStateOut])
def srs_state(user: UserDep, db: DbDep, book_id: int | None = None) -> UnifiedResponse[SrsStateOut]:
    return UnifiedResponse.success(data=_srs(db).get_state(user, book_id))


@router.post("/srs/onboarding", response_model=UnifiedResponse[SrsSettingsOut])
def srs_onboarding(data: OnboardingIn, user: UserDep, db: DbDep) -> UnifiedResponse[SrsSettingsOut]:
    result = _srs(db).save_onboarding(
        user, data.target, data.book_id, data.daily_new_words, data.pronunciation, data.autoplay
    )
    return UnifiedResponse.success(data=result, message="设置已保存")


@router.post("/srs/state", response_model=UnifiedResponse[dict])
def srs_save_state(data: SrsStateIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_srs(db).save_state(user, data.book_id, data.memory, data.session))


@router.post("/srs/complete", response_model=UnifiedResponse[CompleteOut])
def srs_complete(data: CompleteIn, user: UserDep, db: DbDep) -> UnifiedResponse[CompleteOut]:
    result = _srs(db).complete_day(
        user, data.book_id, data.study_date, data.review_count, data.new_count, data.wrong_count
    )
    return UnifiedResponse.success(data=result, message="打卡成功")


@router.post("/srs/sync", response_model=UnifiedResponse[SyncOut])
def srs_sync(data: SyncIn, user: UserDep, db: DbDep) -> UnifiedResponse[SyncOut]:
    return UnifiedResponse.success(data=_srs(db).sync_guest(user, data.book_id, data.memory, data.wordbook))


@router.post("/srs/reset", response_model=UnifiedResponse[dict])
def srs_reset(data: ResetIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_srs(db).reset_book(user, data.book_id), message="已清除学习数据")


@router.get("/srs/stats", response_model=UnifiedResponse[dict])
def srs_stats(user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_srs(db).srs_stats(user))


# ── 单词分类（查看词库 + 按类型背诵） ───────────────────────────────


@router.put("/srs/tag", response_model=UnifiedResponse[dict])
def srs_set_tag(data: SetTagIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_service(db).set_word_tag(user, data.book_id, data.word_id, data.tag))


@router.get("/srs/categories", response_model=UnifiedResponse[list[CategoryOut]])
def srs_categories(user: UserDep, db: DbDep, book_id: int) -> UnifiedResponse[list[CategoryOut]]:
    return UnifiedResponse.success(data=_service(db).get_categories(user, book_id))


# ── 单词测试 ─────────────────────────────────────────────────────────


@router.get("/test/questions", response_model=UnifiedResponse[TestQuestionsOut])
def test_questions(
    user: UserDep,
    db: DbDep,
    book_id: int,
    module: str = "choice",
    question_type: str = "a",
    mode: str = "book",
    count: int = 20,
) -> UnifiedResponse[TestQuestionsOut]:
    result = TestService(db).generate(user, book_id, module, question_type, mode, count)
    return UnifiedResponse.success(data=TestQuestionsOut(**result))


# ── 每日活动埋点 / AI 学习总结 ───────────────────────────────────────


@router.post("/activity", response_model=UnifiedResponse[dict])
def report_activity(data: ActivityIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    from datetime import date, timedelta

    today = date.today()
    d = data.activity_date or today
    if abs((d - today).days) > 1:
        d = today  # 钳制异常时钟
    svc = DailySummaryService(db)
    svc.repo.upsert_activity(user.id, d, **data.model_dump(exclude={"activity_date"}))
    return UnifiedResponse.success(data={"saved": True})


@router.get("/activity/today", response_model=UnifiedResponse[TodayActivityOut])
def activity_today(user: UserDep, db: DbDep) -> UnifiedResponse[TodayActivityOut]:
    return UnifiedResponse.success(data=TodayActivityOut(has_activity=DailySummaryService(db).has_activity_today(user)))


@router.get("/daily-summary", response_model=UnifiedResponse[DailySummaryOut])
def daily_summary(user: UserDep, db: DbDep) -> UnifiedResponse[DailySummaryOut]:
    return UnifiedResponse.success(data=DailySummaryService(db).get_summary(user))


@router.post("/daily-summary/generate", response_model=UnifiedResponse[DailySummaryOut])
async def daily_summary_generate(user: UserDep, db: DbDep) -> UnifiedResponse[DailySummaryOut]:
    result = await DailySummaryService(db).generate(user)
    return UnifiedResponse.success(data=result, message="日报已生成")


# ── 生词本（纯列表） ─────────────────────────────────────────────────


@router.get("/wordbook", response_model=UnifiedResponse[list[WordbookItemOut]])
def list_wordbook(
    user: UserDep, db: DbDep, book_id: int | None = None
) -> UnifiedResponse[list[WordbookItemOut]]:
    return UnifiedResponse.success(data=_srs(db).list_wordbook(user, book_id))


@router.post("/wordbook", response_model=UnifiedResponse[WordbookItemOut])
def add_wordbook(data: WordbookAddIn, user: UserDep, db: DbDep) -> UnifiedResponse[WordbookItemOut]:
    return UnifiedResponse.success(data=_srs(db).add_wordbook(user, data.word_id, data.book_id), message="已加入生词本")


@router.delete("/wordbook", response_model=UnifiedResponse[dict])
def clear_wordbook(user: UserDep, db: DbDep, book_id: int) -> UnifiedResponse[dict]:
    _srs(db).clear_wordbook(user, book_id)
    return UnifiedResponse.success(data={"deleted": True}, message="生词本已清空")


@router.delete("/wordbook/{wb_id}", response_model=UnifiedResponse[dict])
def remove_wordbook(wb_id: int, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    _srs(db).remove_wordbook(user, wb_id)
    return UnifiedResponse.success(data={"deleted": True}, message="已从生词本移除")


# ── 每日一读 ─────────────────────────────────────────────────────────


@router.get("/daily-reading/today", response_model=UnifiedResponse[DailyReadingTodayOut])
async def daily_reading_today(user: UserDep, db: DbDep) -> UnifiedResponse[DailyReadingTodayOut]:
    return UnifiedResponse.success(data=await _reading(db).get_today(user))


@router.put("/daily-reading/level", response_model=UnifiedResponse[dict])
def daily_reading_level(data: ReadingLevelIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_reading(db).set_level_mode(user, data.mode, data.level))


@router.get("/daily-reading/quiz", response_model=UnifiedResponse[dict])
def daily_reading_quiz(user: UserDep, db: DbDep, article_id: int) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_reading(db).get_quiz(user, article_id))


@router.post("/daily-reading/quiz", response_model=UnifiedResponse[dict])
def daily_reading_quiz_submit(data: ReadingQuizSubmitIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_reading(db).submit_quiz(user, data.article_id, data.answers))


@router.post("/daily-reading/words", response_model=UnifiedResponse[dict])
def daily_reading_collect(data: ReadingWordCollectIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_reading(db).collect_word(user, data.article_id, data.word, data.source, data.definition))


@router.put("/daily-reading/words/blacklist", response_model=UnifiedResponse[dict])
def daily_reading_blacklist(data: ReadingBlacklistIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_reading(db).set_blacklist(user, data.word, data.blacklisted))


@router.post("/daily-reading/complete", response_model=UnifiedResponse[dict])
def daily_reading_complete(data: ReadingCompleteIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=_reading(db).complete(user, data.article_id, data.duration_sec), message="今日一读打卡成功")


@router.get("/daily-reading/archive", response_model=UnifiedResponse[list[ReadingArchiveItemOut]])
def daily_reading_archive(user: UserDep, db: DbDep) -> UnifiedResponse[list[ReadingArchiveItemOut]]:
    return UnifiedResponse.success(data=_reading(db).archive(user))


# ── 阅读 ─────────────────────────────────────────────────────────────


@router.get("/reading/articles", response_model=UnifiedResponse[list[ArticleListItemOut]])
def list_articles(
    user: OptionalUserDep, db: DbDep, skip: int = 0, limit: int = 100
) -> UnifiedResponse[list[ArticleListItemOut]]:
    return UnifiedResponse.success(data=_service(db).list_articles(user, skip, limit))


@router.get("/reading/articles/{article_id}", response_model=UnifiedResponse[ArticleOut])
def get_article(article_id: int, user: OptionalUserDep, db: DbDep) -> UnifiedResponse[ArticleOut]:
    return UnifiedResponse.success(data=_service(db).get_article(user, article_id))


@router.get("/reading/notes", response_model=UnifiedResponse[list[NoteOut]])
def list_notes(user: UserDep, db: DbDep) -> UnifiedResponse[list[NoteOut]]:
    return UnifiedResponse.success(data=_service(db).list_notes(user))


@router.post("/reading/notes", response_model=UnifiedResponse[NoteOut])
def create_note(data: NoteCreate, user: UserDep, db: DbDep) -> UnifiedResponse[NoteOut]:
    return UnifiedResponse.success(data=_service(db).create_note(user, data.article_id, data.content), message="笔记已保存")


@router.put("/reading/notes/{note_id}", response_model=UnifiedResponse[NoteOut])
def update_note(note_id: int, data: NoteUpdate, user: UserDep, db: DbDep) -> UnifiedResponse[NoteOut]:
    return UnifiedResponse.success(data=_service(db).update_note(user, note_id, data.content), message="笔记已更新")


@router.delete("/reading/notes/{note_id}", response_model=UnifiedResponse[dict])
def delete_note(note_id: int, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    _service(db).delete_note(user, note_id)
    return UnifiedResponse.success(data={"deleted": True}, message="笔记已删除")


# ── 收藏 ─────────────────────────────────────────────────────────────


@router.get("/collections", response_model=UnifiedResponse[list[CollectionItemOut]])
def list_collections(
    user: UserDep, db: DbDep, item_type: str | None = None
) -> UnifiedResponse[list[CollectionItemOut]]:
    return UnifiedResponse.success(data=_service(db).list_collections(user, item_type))


@router.post("/collections", response_model=UnifiedResponse[CollectionItemOut])
def add_collection(data: CollectionCreate, user: UserDep, db: DbDep) -> UnifiedResponse[CollectionItemOut]:
    return UnifiedResponse.success(data=_service(db).add_collection(user, data.item_type, data.item_id), message="已收藏")


@router.delete("/collections/{collection_id}", response_model=UnifiedResponse[dict])
def remove_collection(collection_id: int, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    _service(db).remove_collection(user, collection_id)
    return UnifiedResponse.success(data={"deleted": True}, message="已取消收藏")


# ── 打卡 / 统计 ──────────────────────────────────────────────────────


@router.post("/checkin", response_model=UnifiedResponse[CheckinStatsOut])
def checkin(user: UserDep, db: DbDep) -> UnifiedResponse[CheckinStatsOut]:
    return UnifiedResponse.success(data=_service(db).checkin(user), message="打卡成功")


@router.get("/checkin/stats", response_model=UnifiedResponse[CheckinStatsOut])
def checkin_stats(user: UserDep, db: DbDep) -> UnifiedResponse[CheckinStatsOut]:
    return UnifiedResponse.success(data=_service(db).checkin_stats(user))


@router.get("/study/stats", response_model=UnifiedResponse[StudyStatsOut])
def study_stats(user: UserDep, db: DbDep) -> UnifiedResponse[StudyStatsOut]:
    return UnifiedResponse.success(data=_service(db).study_stats(user))


# ── 奖励系统 ──────────────────────────────────────────────────────


@router.get("/rewards/overview", response_model=UnifiedResponse[RewardOverviewOut])
def rewards_overview(user: UserDep, db: DbDep) -> UnifiedResponse[RewardOverviewOut]:
    return UnifiedResponse.success(data=RewardService(db).overview(user))


@router.get("/rewards/shop", response_model=UnifiedResponse[list[ShopItemOut]])
def rewards_shop(user: UserDep, db: DbDep) -> UnifiedResponse[list[ShopItemOut]]:
    return UnifiedResponse.success(data=RewardService(db).shop(user))


@router.post("/rewards/collect", response_model=UnifiedResponse[CollectOut])
def rewards_collect(user: UserDep, db: DbDep) -> UnifiedResponse[CollectOut]:
    return UnifiedResponse.success(data=RewardService(db).collect(user), message="奖励结算完成")


@router.post("/rewards/redeem", response_model=UnifiedResponse[ShopItemOut])
def rewards_redeem(data: RedeemIn, user: UserDep, db: DbDep) -> UnifiedResponse[ShopItemOut]:
    return UnifiedResponse.success(data=RewardService(db).redeem(user, data.item_key), message="兑换成功")


@router.post("/rewards/equip", response_model=UnifiedResponse[dict])
def rewards_equip(data: EquipIn, user: UserDep, db: DbDep) -> UnifiedResponse[dict]:
    return UnifiedResponse.success(data=RewardService(db).equip(user, data.item_key))
